From d31effb4277bd557f5ccf16d909422718c1e49d0 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke@hauke-m.de>
Date: Mon, 13 Oct 2025 23:59:33 +0200
Subject: [PATCH] ubusd: Fix out of bounds access in event register message

The code assumes that the provided pattern is at least one byte long.
reject shorter patterns.

Empty messages could lead to heap corruptions and ubusd_acl_check()
bypass.

Reported-by: Karsten Sperling <ksperling@apple.com>
Fixes: 12623b43060a ("trim the wildcard of partial patterns to keep the avl tree sorted properly")
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
CVE: CVE-2025-62526
(cherry picked from commit d31effb4277bd557f5ccf16d909422718c1e49d0)
[Backport to ubus 2022-02-21-b32a0e17 (OpenWrt 19.07, Gluon v2021.1.x) by
 Freifunk im Neanderland <projekt@neanderfunk.de>: code unchanged,
 hunks refreshed for line offsets only. Part of the CVE-2025-62526 fix
 set (d31effb, aa4a7ee, 60e0404), as in OpenWrt 24.10 a790196993.]
---
 ubusd_event.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/ubusd_event.c b/ubusd_event.c
--- a/ubusd_event.c
+++ b/ubusd_event.c
@@ -81,6 +81,9 @@ static int ubusd_alloc_event_pattern(struct ubus_client *cl, struct blob_attr *m
 	pattern = blobmsg_data(attr[EVREG_PATTERN]);
 
 	len = strlen(pattern);
+	if (len <= 0)
+		return UBUS_STATUS_PERMISSION_DENIED;
+
 	if (pattern[len - 1] == '*') {
 		partial = true;
 		pattern[len - 1] = 0;
